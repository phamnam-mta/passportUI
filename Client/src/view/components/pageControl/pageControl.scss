.page-control {
    display: flex;
    gap: 5px;
    align-items: center;

    .page-number {
        display: flex;
        gap: 5px;
        align-items: center;
    }

    /* Chrome, Safari, Edge, Opera */
    input::-webkit-outer-spin-button,
    input::-webkit-inner-spin-button {
        -webkit-appearance: none;
        margin: 0;
    }

    /* Firefox */
    input[type="number"] {
        -moz-appearance: textfield;
    }
}
