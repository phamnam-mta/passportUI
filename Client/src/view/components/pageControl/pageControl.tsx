import * as React from "react";
import { TextField } from "@fluentui/react";
import { DefaultIconButton } from "view/components/buttons/buttons";

import "./pageControl.scss";

interface IPageControlProps {
    disabled: boolean;
    currentPage: number | undefined;
    numPages: number | undefined;
    onPageChange: (page: number) => void;
    onPreviousClick: () => void;
    onNextClick: () => void;
}

interface IPageControlState {
    displayPage: number;
}

export class PageControl extends React.PureComponent<IPageControlProps, IPageControlState> {
    private inputRef;
    private inputTimer;
    private INPUT_CHANGE_DELAY = 400;

    constructor(props) {
        super(props);
        this.inputRef = React.createRef();
        this.state = {
            displayPage: props.currentPage,
        };
    }

    public componentDidUpdate(prevProps) {
        const { currentPage } = this.props;
        if (prevProps.currentPage !== currentPage && currentPage) {
            this.setState({ displayPage: currentPage });
        }
    }

    private getDisplayPage = () => {
        const { displayPage } = this.state;
        return displayPage && displayPage > 0 ? displayPage.toString() : "";
    };

    private getNumPageString = () => {
        const { numPages } = this.props;
        const preposition = "of";
        return numPages ? `${preposition} ` + numPages : `${preposition} ##`;
    };

    private isPageInputDisabled = () => {
        const { numPages } = this.props;
        return numPages === 1 || !this.getDisplayPage();
    };

    private isPreviousDisabled = () => {
        return this.state.displayPage === 1 || !this.getDisplayPage();
    };

    private isNextDisabled = () => {
        return this.state.displayPage === this.props.numPages || !this.getDisplayPage();
    };

    private delayPageChange = (page: number) => {
        if (this.inputTimer) {
            // Reset timer.
            clearTimeout(this.inputTimer);
        }
        // Set timer to delay dispatch page change action.
        this.inputTimer = setTimeout(() => {
            this.props.onPageChange(page);
        }, this.INPUT_CHANGE_DELAY);
    };

    private onDisplayPageChange = (event, newValue?: string) => {
        const { numPages } = this.props;
        const page = newValue ? parseInt(newValue) : 0;
        if (newValue && numPages && page > 0 && page <= numPages) {
            this.setState({ displayPage: page });
            this.delayPageChange(page);
        }
    };

    private onPageEnter = (event) => {
        if (event.key === "Enter") {
            this.inputRef.current.blur();
        }
    };

    public render() {
        const { disabled, onPreviousClick, onNextClick } = this.props;
        return (
            <div className="page-control">
                <DefaultIconButton
                    title="Previous page"
                    name={"ChevronLeft"}
                    disabled={disabled || this.isPreviousDisabled()}
                    onClick={onPreviousClick}
                />
                <div className="page-number">
                    <div className="current-page">
                        <TextField
                            componentRef={this.inputRef}
                            disabled={disabled || this.isPageInputDisabled()}
                            type="number"
                            styles={{ root: { width: 60 } }}
                            value={this.getDisplayPage()}
                            onChange={this.onDisplayPageChange}
                            onKeyPress={this.onPageEnter}
                            ariaLabel="Current page"
                        />
                    </div>
                    <div className="total-page">{this.getNumPageString()}</div>
                </div>
                <DefaultIconButton
                    name={"ChevronRight"}
                    title="Next page"
                    disabled={disabled || this.isNextDisabled()}
                    onClick={onNextClick}
                />
            </div>
        );
    }
}

export default PageControl;
