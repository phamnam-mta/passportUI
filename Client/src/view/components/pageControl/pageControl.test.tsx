import * as React from "react";
import { shallow } from "enzyme";
import { PageControl } from "./pageControl";

describe("<PageControl />", () => {
    let baseProps;

    beforeEach(() => {
        jest.useFakeTimers();
        baseProps = {
            disabled: false,
            currentPage: 3,
            numPages: 10,
            onPageChange: jest.fn(),
            onPreviousClick: jest.fn(),
            onNextClick: jest.fn(),
        };
    });

    describe("Rendering", () => {
        it("should render correctly", () => {
            const wrapper = shallow(<PageControl {...baseProps} />);

            expect(wrapper).toMatchSnapshot();
        });

        it("should match snapshot, when page control is disabled", () => {
            const props = { ...baseProps, disabled: true };

            const wrapper = shallow(<PageControl {...props} />);

            expect(wrapper).toMatchSnapshot();
        });

        it("should match snapshot, when current page is the first page", () => {
            const props = { ...baseProps, currentPage: 1 };

            const wrapper = shallow(<PageControl {...props} />);

            expect(wrapper).toMatchSnapshot();
        });

        it("should match snapshot, when current page is the last page", () => {
            const props = { ...baseProps, currentPage: 10 };

            const wrapper = shallow(<PageControl {...props} />);

            expect(wrapper).toMatchSnapshot();
        });

        it("should match snapshot, when number of pages is 1", () => {
            const props = { ...baseProps, currentPage: 1, numPages: 1 };

            const wrapper = shallow(<PageControl {...props} />);

            expect(wrapper).toMatchSnapshot();
        });

        it("should match snapshot, when there's no current page and number of pages", () => {
            const props = { ...baseProps, currentPage: undefined, numPages: undefined };

            const wrapper = shallow(<PageControl {...props} />);

            expect(wrapper).toMatchSnapshot();
        });
    });

    describe("Event Handling", () => {
        const textFieldSelector = "StyledTextFieldBase";

        it("should set display page, when current page is changed", () => {
            const wrapper = shallow(<PageControl {...baseProps} />);
            wrapper.setProps({ currentPage: 4 });

            expect(wrapper.state("displayPage")).toBe(4);
        });

        it("should handle previous page, when the previous button is clicked", () => {
            const wrapper = shallow(<PageControl {...baseProps} />);
            const previousButton = wrapper.find("DefaultIconButton").first();
            previousButton.simulate("click");

            expect(baseProps.onPreviousClick).toBeCalledTimes(1);
        });

        it("should handle previous page, when the next button is clicked", () => {
            const wrapper = shallow(<PageControl {...baseProps} />);
            const previousButton = wrapper.find("DefaultIconButton").last();
            previousButton.simulate("click");

            expect(baseProps.onNextClick).toBeCalledTimes(1);
        });

        it("should handle display page change, when the page input is changed", () => {
            const page = 5;

            const wrapper = shallow(<PageControl {...baseProps} />);
            const textField = wrapper.find(textFieldSelector) as any;

            // Normal.
            textField.simulate("change", {}, page.toString());
            expect(wrapper.state("displayPage")).toBe(page);

            // Do not change the display page when out of bound.
            textField.simulate("change", {}, "12");
            expect(wrapper.state("displayPage")).toBe(page);

            textField.simulate("change", {}, "0");
            expect(wrapper.state("displayPage")).toBe(page);

            textField.simulate("change", {}, undefined);
            expect(wrapper.state("displayPage")).toBe(page);
        });

        it("should trigger onPageChange with delay, when the page input is changed continuously", () => {
            const wrapper = shallow(<PageControl {...baseProps} />);
            const textField = wrapper.find(textFieldSelector) as any;
            const clearTimeoutSpy = jest.spyOn(global, "clearTimeout");
            const setTimeoutSpy = jest.spyOn(global, "setTimeout");

            textField.simulate("change", {}, "5");
            textField.simulate("change", {}, "6");
            textField.simulate("change", {}, "7");
            jest.runAllTimers();

            expect(clearTimeoutSpy).toBeCalledTimes(2);
            expect(setTimeoutSpy).toBeCalledTimes(3);
            expect(baseProps.onPageChange).toBeCalledTimes(1);
            expect(baseProps.onPageChange).toBeCalledWith(7);
        });

        it("should handle blur text field, when a key is pressed", () => {
            const wrapper = shallow(<PageControl {...baseProps} />);
            wrapper.instance()["inputRef"] = { current: { blur: jest.fn() } };
            const textField = wrapper.find(textFieldSelector) as any;

            // Press Enter.
            textField.simulate("keypress", { key: "Enter" });
            expect(wrapper.instance()["inputRef"].current.blur).toBeCalled();

            // Press other key should not trigger blur.
            wrapper.instance()["inputRef"] = { current: { blur: jest.fn() } };
            textField.simulate("keypress", { key: "Alt" });
            expect(wrapper.instance()["inputRef"].current.blur).not.toBeCalled();
        });
    });
});
