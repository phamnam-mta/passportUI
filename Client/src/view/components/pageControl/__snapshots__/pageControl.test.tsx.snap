// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageControl /> Rendering should match snapshot, when current page is the first page 1`] = `
<div
  className="page-control"
>
  <DefaultIconButton
    disabled={true}
    name="ChevronLeft"
    onClick={[MockFunction]}
    title="Previous page"
  />
  <div
    className="page-number"
  >
    <div
      className="current-page"
    >
      <StyledTextFieldBase
        ariaLabel="Current page"
        componentRef={
          Object {
            "current": null,
          }
        }
        disabled={false}
        onChange={[Function]}
        onKeyPress={[Function]}
        styles={
          Object {
            "root": Object {
              "width": 60,
            },
          }
        }
        type="number"
        value="1"
      />
    </div>
    <div
      className="total-page"
    >
      of 10
    </div>
  </div>
  <DefaultIconButton
    disabled={false}
    name="ChevronRight"
    onClick={[MockFunction]}
    title="Next page"
  />
</div>
`;

exports[`<PageControl /> Rendering should match snapshot, when current page is the last page 1`] = `
<div
  className="page-control"
>
  <DefaultIconButton
    disabled={false}
    name="ChevronLeft"
    onClick={[MockFunction]}
    title="Previous page"
  />
  <div
    className="page-number"
  >
    <div
      className="current-page"
    >
      <StyledTextFieldBase
        ariaLabel="Current page"
        componentRef={
          Object {
            "current": null,
          }
        }
        disabled={false}
        onChange={[Function]}
        onKeyPress={[Function]}
        styles={
          Object {
            "root": Object {
              "width": 60,
            },
          }
        }
        type="number"
        value="10"
      />
    </div>
    <div
      className="total-page"
    >
      of 10
    </div>
  </div>
  <DefaultIconButton
    disabled={true}
    name="ChevronRight"
    onClick={[MockFunction]}
    title="Next page"
  />
</div>
`;

exports[`<PageControl /> Rendering should match snapshot, when number of pages is 1 1`] = `
<div
  className="page-control"
>
  <DefaultIconButton
    disabled={true}
    name="ChevronLeft"
    onClick={[MockFunction]}
    title="Previous page"
  />
  <div
    className="page-number"
  >
    <div
      className="current-page"
    >
      <StyledTextFieldBase
        ariaLabel="Current page"
        componentRef={
          Object {
            "current": null,
          }
        }
        disabled={true}
        onChange={[Function]}
        onKeyPress={[Function]}
        styles={
          Object {
            "root": Object {
              "width": 60,
            },
          }
        }
        type="number"
        value="1"
      />
    </div>
    <div
      className="total-page"
    >
      of 1
    </div>
  </div>
  <DefaultIconButton
    disabled={true}
    name="ChevronRight"
    onClick={[MockFunction]}
    title="Next page"
  />
</div>
`;

exports[`<PageControl /> Rendering should match snapshot, when page control is disabled 1`] = `
<div
  className="page-control"
>
  <DefaultIconButton
    disabled={true}
    name="ChevronLeft"
    onClick={[MockFunction]}
    title="Previous page"
  />
  <div
    className="page-number"
  >
    <div
      className="current-page"
    >
      <StyledTextFieldBase
        ariaLabel="Current page"
        componentRef={
          Object {
            "current": null,
          }
        }
        disabled={true}
        onChange={[Function]}
        onKeyPress={[Function]}
        styles={
          Object {
            "root": Object {
              "width": 60,
            },
          }
        }
        type="number"
        value="3"
      />
    </div>
    <div
      className="total-page"
    >
      of 10
    </div>
  </div>
  <DefaultIconButton
    disabled={true}
    name="ChevronRight"
    onClick={[MockFunction]}
    title="Next page"
  />
</div>
`;

exports[`<PageControl /> Rendering should match snapshot, when there's no current page and number of pages 1`] = `
<div
  className="page-control"
>
  <DefaultIconButton
    disabled={true}
    name="ChevronLeft"
    onClick={[MockFunction]}
    title="Previous page"
  />
  <div
    className="page-number"
  >
    <div
      className="current-page"
    >
      <StyledTextFieldBase
        ariaLabel="Current page"
        componentRef={
          Object {
            "current": null,
          }
        }
        disabled={true}
        onChange={[Function]}
        onKeyPress={[Function]}
        styles={
          Object {
            "root": Object {
              "width": 60,
            },
          }
        }
        type="number"
        value=""
      />
    </div>
    <div
      className="total-page"
    >
      of ##
    </div>
  </div>
  <DefaultIconButton
    disabled={true}
    name="ChevronRight"
    onClick={[MockFunction]}
    title="Next page"
  />
</div>
`;

exports[`<PageControl /> Rendering should render correctly 1`] = `
<div
  className="page-control"
>
  <DefaultIconButton
    disabled={false}
    name="ChevronLeft"
    onClick={[MockFunction]}
    title="Previous page"
  />
  <div
    className="page-number"
  >
    <div
      className="current-page"
    >
      <StyledTextFieldBase
        ariaLabel="Current page"
        componentRef={
          Object {
            "current": null,
          }
        }
        disabled={false}
        onChange={[Function]}
        onKeyPress={[Function]}
        styles={
          Object {
            "root": Object {
              "width": 60,
            },
          }
        }
        type="number"
        value="3"
      />
    </div>
    <div
      className="total-page"
    >
      of 10
    </div>
  </div>
  <DefaultIconButton
    disabled={false}
    name="ChevronRight"
    onClick={[MockFunction]}
    title="Next page"
  />
</div>
`;
